infrastructure/src/main/java/org/corfudb/infrastructure/RemoteMonitoringService.java

```diff
@@ -421,25 +423,45 @@ private synchronized void runDetectionTasks() {
      * @return Detector task.
      */
     private DetectorTask restoreRedundancyAndMergeSegments(Layout layout) {
-        int segmentsCount = layout.getSegments().size();
+        String localEndpoint = serverContext.getLocalEndpoint();
 
-        if (segmentsCount == 1) {
-            log.debug("No segments to merge. Skipping step.");
-            return DetectorTask.SKIPPED;
-        } else if (!mergeSegmentsTask.isDone()) {
-            log.debug("Merge segments task already in progress. Skipping spawning another task.");
+        // Check that the task is not currently running.
+        if(!mergeSegmentsTask.isDone()){
+            log.trace("Merge segments task already in progress. Skipping spawning another task.");
             return DetectorTask.SKIPPED;
         }
 
-        log.debug("Number of segments present: {}. Spawning task to merge segments.", segmentsCount);
+        // Check that the current node can invoke a restoration action,
+        // also verify that the current node is healed (not present in the unresponsive list).
+        if (RedundancyCalculator.canRestoreRedundancyOrMergeSegments(layout, localEndpoint) &&
+                !layout.getUnresponsiveServers().contains(localEndpoint)) {
+            log.info("Layout requires restoration: {}. Spawning task to merge segments on {}.",
+                    layout, localEndpoint);
+            Supplier<Boolean> redundancyAction = () -> handleMergeSegments(
+                    localEndpoint, runtimeSingletonResource, layout, MERGE_SEGMENTS_RETRY_QUERY_TIMEOUT
+            );
+            mergeSegmentsTask = CompletableFuture.supplyAsync(redundancyAction, failureDetectorWorker);
```

Comment:
Why redispatch if you can just use that thread directly ?
It seems like you have a future just to synchronize and no-op the second invocation.

Reply:
This code is currently present in the master, and I did not really touch it.
https://github.com/CorfuDB/CorfuDB/blob/ba0b92ce51b565052d716a28ea033f57a039af87/infrastructure/src/main/java/org/corfudb/infrastructure/RemoteMonitoringService.java#L440